radar-upload-backend/src/main/java/org/radarbase/upload/service/managementportal/MPProjectService.kt

```diff
@@ -56,11 +57,15 @@ class MPProjectService(@Context private val mpClient: MPClient): UploadProjectSe
 
     override fun projectUsers(projectId: String): List<User> {
         val projectParticipants = participants.computeIfAbsent(projectId) {
-            CachedSet(Duration.ofMinutes(30), Duration.ofMinutes(1)) {
+            CachedSet(Duration.ofMinutes(config.syncParticipants!!), Duration.ofMinutes(1)) {
                 mpClient.readParticipants(projectId)
             }
         }
 
         return projectParticipants.get().toList()
     }
+
+    override fun userByExternalId(projectId: String, externalUserId: String): User? =
+            projectUsers(projectId).firstOrNull { it.externalId == externalUserId }
```

Comment:
I would suggest to do a `find` here instead, and throw an `IllegalStateException` if there is more than one user with the same externalUserId. Alternatively, make this method return a List<User> and throw an exception in `RecordMapperImpl.resolveUserId`

Just to check: how does ManagementPortal handle duplicate external IDs? Is it forbidden?

Reply:
ManagmentPortal does not allow duplicate externalIds within a project.